### PR TITLE
Convert exception help context parsing to managed

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Allows removing some native string manipulation code.

The code path looks only used for COM or OLEAut marshaling. I don't think Unicode categorization of whitespace would make any real difference here.

Potential OOM is same with previous code. No other exception thrown.

### CODE REVIEW

**Comment by jkotas** [comment]
Is this exercised by any tests?

**Comment by huoyaoyuan** [comment]
I did not find for a test. I'm not even sure if this code is alive. The field ends up get used here:
https://github.com/dotnet/runtime/blob/cb3bd225a88dd75ad804a3701a2ac9b2553a4973/src/coreclr/vm/interoputil.cpp#L182-L214
Looks like a thing for COM interop.

**Comment by jkotas** [comment]
Yeah, that is our pain point with the legacy COM support. We do not have a good test coverage for it and every time we touch something in it we end up with very hard to track down regression a few months later.

**Comment by huoyaoyuan** [comment]
What would be the test approach then? I didn't even know about the help context ID before exploring this.

**Comment by AaronRobinsonMSFT** [comment]
This is an area that is going to need testing of some sort. The way this can likely be validated is by setting a thread local `IErrorInfo` in a native COM server via [`SetErrorInfo`](https://docs.microsoft.com/windows/win32/api/oleauto/nf-oleauto-seterrorinfo). See the various combinations under https://github.com/dotnet/runtime/tree/main/src/tests/Interop/COM.

**Comment by huoyaoyuan** [comment]
I have totally no confident whether the code is correct about ref counting. Today I learned the strong desire to create a COM Plus back into those days.

File: src/tests/Interop/COM/ServerContracts/Server.Contracts.h [hunk 1]
```diff
@@ -375,6 +375,10 @@ IErrorMarshalTesting : IUnknown
         /*[in]*/ int hresultToReturn ) = 0;
       virtual int STDMETHODCALLTYPE Return_As_HResult_Struct (
         /*[in]*/ int hresultToReturn ) = 0;
+      virtual HRESULT STDMETHODCALLTYPE Throw_HResult_HelpLink (
+        /*[in]*/ int hresultToReturn,
+        /*[in]*/ BSTR helpLink,
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/ServerContracts/Server.Contracts.h:380]: This should be `LPCWSTR`.

File: src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h [hunk 2]
```diff
@@ -26,6 +26,25 @@ class ErrorMarshalTesting : public UnknownImpl, public IErrorMarshalTesting
         return hresultToReturn;
     }
 
+    DEF_FUNC(Throw_HResult_HelpLink)(
+        /*[in]*/ int hresultToReturn,
+        /*[in]*/ BSTR helpLink,
+        /*[in]*/ DWORD helpContext)
+    {
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h:33]: Declare a local `HRESULT hr;`

File: src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h [hunk 3]
```diff
@@ -26,6 +26,25 @@ class ErrorMarshalTesting : public UnknownImpl, public IErrorMarshalTesting
...
+        ICreateErrorInfo* pCreateErrInfo;
+        CreateErrorInfo(&pCreateErrInfo);
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h:35]:
```suggestion
        RETURN_IF_FAILED(::CreateErrorInfo(&pCreateErrInfo));
```

File: src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h [hunk 4]
```diff
@@ -26,6 +26,25 @@ class ErrorMarshalTesting : public UnknownImpl, public IErrorMarshalTesting
...
+        pCreateErrInfo->SetHelpFile(helpLink);
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h:36]:
```suggestion
        RETURN_IF_FAILED(pCreateErrInfo->SetHelpFile(helpLink));
```

File: src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h [hunk 5]
```diff
@@ -26,6 +26,25 @@ class ErrorMarshalTesting : public UnknownImpl, public IErrorMarshalTesting
...
+        pCreateErrInfo->SetHelpContext(helpContext);
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h:37]:
```suggestion
        RETURN_IF_FAILED(pCreateErrInfo->SetHelpContext(helpContext));
```

File: src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h [hunk 6]
```diff
@@ -26,6 +26,25 @@ class ErrorMarshalTesting : public UnknownImpl, public IErrorMarshalTesting
...
+
+        IErrorInfo* pErrInfo;
+        pCreateErrInfo->QueryInterface(IID_IErrorInfo, (void**)&pErrInfo);
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h:40]:
```suggestion
        RETURN_IF_FAILED(pCreateErrInfo->QueryInterface(IID_IErrorInfo, (void**)&pErrInfo));
```

File: src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h [hunk 7]
```diff
@@ -26,6 +26,25 @@ class ErrorMarshalTesting : public UnknownImpl, public IErrorMarshalTesting
...
+        SetErrorInfo(0, pErrInfo);
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h:41]:
```suggestion
        RETURN_IF_FAILED(SetErrorInfo(0, pErrInfo));
```

File: src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h [hunk 8]
```diff
@@ -26,6 +26,25 @@ class ErrorMarshalTesting : public UnknownImpl, public IErrorMarshalTesting
...
+        pErrInfo->Release();
+        pCreateErrInfo->Release();
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h:43]: Both of these releases can be removed.

File: src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h [hunk 9]
```diff
@@ -26,6 +26,25 @@ class ErrorMarshalTesting : public UnknownImpl, public IErrorMarshalTesting
         return hresultToReturn;
     }
 
+    DEF_FUNC(Throw_HResult_HelpLink)(
+        /*[in]*/ int hresultToReturn,
+        /*[in]*/ BSTR helpLink,
+        /*[in]*/ DWORD helpContext)
+    {
+        ICreateErrorInfo* pCreateErrInfo;
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h:34]:
```suggestion
        ComSmartPtr<ICreateErrorInfo> pCreateErrInfo;
```

File: src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h [hunk 10]
```diff
@@ -26,6 +26,25 @@ class ErrorMarshalTesting : public UnknownImpl, public IErrorMarshalTesting
...
+        CreateErrorInfo(&pCreateErrInfo);
+        pCreateErrInfo->SetHelpFile(helpLink);
+        pCreateErrInfo->SetHelpContext(helpContext);
+
+        IErrorInfo* pErrInfo;
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h:39]:
```suggestion
        ComSmartPtr<IErrorInfo> pErrInfo;
```

File: src/tests/Interop/COM/NativeClients/Primitives/ErrorTests.cpp [hunk 11]
```diff
@@ -75,6 +75,48 @@ namespace
...
+            E_POINTER,
+            E_ACCESSDENIED,
+            E_INVALIDARG,
+            E_UNEXPECTED,
+            HRESULT{-1},
+            S_FALSE,
+            HRESULT{2}
+        };
+
+        BSTR helpLink = SysAllocString(OLESTR("C:\\Windows\\system32\\dummy.hlp"));
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NativeClients/Primitives/ErrorTests.cpp:95]:
```suggestion
        BSTR helpLink = SysAllocString(L"X:\\NotA\\RealPath\\dummy.hlp");
```

File: src/tests/Interop/COM/NativeClients/Primitives/ErrorTests.cpp [hunk 12]
```diff
@@ -75,6 +75,48 @@ namespace
...
+
+        for (int i = 0; i < ARRAY_SIZE(hrs); ++i)
+        {
+            HRESULT hr = hrs[i];
+            DWORD helpContext = (DWORD)(i + 0x1234);
+            HRESULT hrMaybe = et->Throw_HResult_HelpLink(hr, helpLink, helpContext);
+            THROW_FAIL_IF_FALSE(hr == hrMaybe);
+            
+            IErrorInfo* pErrInfo;
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NativeClients/Primitives/ErrorTests.cpp:104]:
```suggestion
            ComSmartPtr<IErrorInfo> pErrInfo;
```

File: src/tests/Interop/COM/NativeClients/Primitives/ErrorTests.cpp [hunk 13]
```diff
@@ -75,6 +75,48 @@ namespace
...
+
+            BSTR helpLinkMaybe;
+            THROW_IF_FAILED(pErrInfo->GetHelpFile(&helpLinkMaybe));
+            THROW_FAIL_IF_FALSE(VarBstrCmp(helpLink, helpLinkMaybe, LANG_ENGLISH, 0) == VARCMP_EQ);
+            SysFreeString(helpLinkMaybe);
+
+            DWORD helpContextMaybe;
+            THROW_IF_FAILED(pErrInfo->GetHelpContext(&helpContextMaybe));
+            THROW_FAIL_IF_FALSE(helpContext == helpContextMaybe);
+            pErrInfo->Release();
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NativeClients/Primitives/ErrorTests.cpp:115]: Remove

File: src/tests/Interop/COM/NativeClients/Primitives/ErrorTests.cpp [hunk 14]
```diff
@@ -75,6 +75,48 @@ namespace
...
+            DWORD helpContext = (DWORD)(i + 0x1234);
+            HRESULT hrMaybe = et->Throw_HResult_HelpLink(hr, helpLink, helpContext);
+            THROW_FAIL_IF_FALSE(hr == hrMaybe);
+            
+            IErrorInfo* pErrInfo;
+            THROW_IF_FAILED(GetErrorInfo(0, &pErrInfo));
+
+            BSTR helpLinkMaybe;
+            THROW_IF_FAILED(pErrInfo->GetHelpFile(&helpLinkMaybe));
+            THROW_FAIL_IF_FALSE(VarBstrCmp(helpLink, helpLinkMaybe, LANG_ENGLISH, 0) == VARCMP_EQ);
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NativeClients/Primitives/ErrorTests.cpp:109]: I'd avoid `VarBstrCmp`. Instead use `TP_wcmp_s`

File: src/tests/Interop/COM/NETClients/Primitives/ErrorTests.cs [hunk 15]
```diff
@@ -40,21 +41,29 @@ private void VerifyReturnHResult()
...
             foreach (var hr in hrs)
             {
                 Assert.Equal(hr, this.server.Return_As_HResult(hr));
                 Assert.Equal(hr, this.server.Return_As_HResult_Struct(hr).hr);
             }
         }
+
+        private void VerifyHelpLink()
+        {
+            string helpLink = "C:\\Windows\\system32\\dummy.hlp";
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NETClients/Primitives/ErrorTests.cs:63]:
```suggestion
            string helpLink = "X:\\NotA\\RealPath\\dummy.hlp";
```

**Comment by huoyaoyuan** [comment]
`Compiler optimization assumption invalid: EE expects method to exist: System.Exception:GetHelpContext`

What's missing for this? Is the metasig definition correct?

**Comment by AaronRobinsonMSFT** [comment]
> `Compiler optimization assumption invalid: EE expects method to exist: System.Exception:GetHelpContext`
> 
> What's missing for this? Is the metasig definition correct?

The code looks correct. Where are you seeing this failure? CI failures appear to be for testing.

**Comment by jkotas** [comment]
This is due to interaction between corelib.h and the linker. The ifdefs in corelib.h do not "just work". Check what's done for other similar COM-specific methods in corelib.h.

**Comment by huoyaoyuan** [comment]
It fails for native client testing. I've removed the ifdef now.
Going to debug it locally. But it seems that there isn't such native->managed marshaling, which is out of scope of this PR.

**Comment by huoyaoyuan** [comment]
Debugging mixed managed/unmanaged code with interop and runtime generated stubs is ... a brand new experience.
I stepped with disassembly and then found the code calling `ISupportsErrorInfo`.
Both sides of tests are now passing locally.

**Comment by huoyaoyuan** [comment]
CI failures looks unrelated? NativeAOT has failed for many times recently.

File: src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h [hunk 16]
```diff
@@ -26,12 +26,38 @@ class ErrorMarshalTesting : public UnknownImpl, public IErrorMarshalTesting
...
+        RETURN_IF_FAILED(pCreateErrInfo->QueryInterface(IID_IErrorInfo, (void**)&pErrInfo));
+        RETURN_IF_FAILED(SetErrorInfo(0, pErrInfo));
+
+        return HRESULT{ hresultToReturn };
+    }
+
+    DEF_FUNC(InterfaceSupportsErrorInfo)(
+        /* [in] */ __RPC__in REFIID riid)
+    {
+        return S_OK;
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/COM/NativeServer/ErrorMarshalTesting.h:52]: This should only return `S_OK` when the interface queried supports it. In this case, I think that is `IErrorMarshalTesting`.

File: src/coreclr/vm/corelib.h [hunk 17]
```diff
@@ -330,11 +330,12 @@ DEFINE_FIELD_U(_xptrs,             ExceptionObject,    _xptrs)
 DEFINE_FIELD_U(_xcode,             ExceptionObject,    _xcode)
 DEFINE_FIELD_U(_HResult,           ExceptionObject,    _HResult)
 DEFINE_CLASS(EXCEPTION,             System,                 Exception)
+DEFINE_METHOD(EXCEPTION,            INTERNAL_PRESERVE_STACK_TRACE, InternalPreserveStackTrace, IM_RetVoid)
+// Following Exception members are only used when FEATURE_COMINTEROP
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/corelib.h:334]: Instead of this comment can we use the `#ifdef FEATURE_COMINTEROP` mechanism?

**Review by AaronRobinsonMSFT — APPROVED** [review]
A few nits, but overall good. Thanks!

File: src/coreclr/vm/corelib.h (hunk 17 above)
Comment (huoyaoyuan) [reply src/coreclr/vm/corelib.h:334]: The comment in this file says that ILLink can only process the markers at type level:
https://github.com/dotnet/runtime/blob/20e5237104ddd178f12bc501352e45a5e91e5048/src/coreclr/vm/corelib.h#L71-L79

**Comment by huoyaoyuan** [comment]
The NativeAOT failure looks different between commits.

**Comment by AaronRobinsonMSFT** [comment]
Failure looks like https://github.com/dotnet/runtime/issues/69192.

File: src/coreclr/System.Private.CoreLib/src/System/Exception.CoreCLR.cs [hunk 18]
```diff
@@ -274,5 +274,32 @@ private bool CanSetRemoteStackTrace()
...
+                return helpFile;
+            }
+
+            for (digitEnd = poundPos + 1; digitEnd < helpFile.Length; digitEnd++)
+            {
+                if (char.IsWhiteSpace(helpFile[digitEnd]))
+                    break;
+            }
+
+            if (uint.TryParse(helpFile.AsSpan(poundPos + 1, digitEnd - poundPos - 1), out helpContext))
```
Comment (GSPP) [inline src/coreclr/System.Private.CoreLib/src/System/Exception.CoreCLR.cs:297]: Should this not be culture invariant?

Comment (huoyaoyuan) [reply src/coreclr/System.Private.CoreLib/src/System/Exception.CoreCLR.cs:297]: AFAIK the default format of uint32 isn't affected by culture at all. Negative sign, digit separator and decimal point are not allowed.

**Pull request merged**